src/compile/scale.ts

```diff
@@ -43,8 +51,46 @@ export function compileScales(channels: Channel[], model: Model) {
         }
       });
 
-      return scaleDef;
-    });
+      // add additional scales needed to support ordinal color scales also for legends
+      if (channel === COLOR && fieldDef.legend && (fieldDef.type === ORDINAL || fieldDef.bin || fieldDef.timeUnit)) {
+        scales.push({
+          name: INVERSE_RANK,
+          type: ORDINAL,
+          domain: {
+            data: model.dataTable(),
+            // use rank_<field> for ordinal type, for bin and timeUnit use default field
+            field: model.field(COLOR, (fieldDef.bin || fieldDef.timeUnit) ? {} : {prefn: 'rank_'}), sort: true
+          },
+          range: {data: model.dataTable(), field: model.field(COLOR), sort: true}
```

Comment:
I admit that the name is misleading and I can change it to identity for bin and time unit. However, i'd love to know how you add legends without an identity scale. The code would get more complicated if I had separate cases for ordinal and bin/time unit but I could rename INVERSE_RANK to INVERSE to make it less confusing.